### ISSUE TITLE
All day event dates incorrect in calendar ListItems created in Graph API

### ISSUE THREAD

**Issue by mikepowell** [SharePoint/sp-dev-docs#2755]
#### Category
- [ ] Question
- [ ] Typo
- [x] Bug
- [ ] Additional article idea

#### Expected or Desired Behavior

When creating an all-day calendar ListItem, the start time should be 00:00 on the start date, and the end time should be 23:59 on the end date.

#### Observed Behavior

The resulting ListItem starts at 00:00 on the day *before* the expected start date. The end date and time are as expected.

#### Steps to Reproduce

```
POST https://graph.microsoft.com/v1.0/sites/root:/{sitename}:/lists/{listname}/items
```
```js
// Single day, all-day event on 8/31/2018
{
    "fields": {
        "Title": "Test Event",
        "EventDate": "2018-08-31T00:00:00",
        "EndDate": "2018-08-31T23:59:00",
        "fAllDayEvent": true
    }
}
```

#### Result

```js
"fields": {
    "EventDate": "2018-08-30T00:00:00Z",  // A day earlier than expected
    "EndDate": "2018-08-31T23:59:00Z",
    ...
}
```

Note that the results are the same if EventDate and EndDate are specified with no time information, e.g. "2018-08-31".

#### Workaround

Adding a day to the EventDate field causes the created item to be correct. This works even if creating a single-day event, where it's necessary to specify a start date that's _later than_ the end date:

```js
{
    "fields": {
        "Title": "Test Event",
        "EventDate": "2018-09-01T00:00:00",  // Add a day.
        "EndDate": "2018-08-31T23:59:00", 
        "fAllDayEvent": true
    }
}
```

```js
"fields": {
    "EventDate": "2018-08-31T00:00:00Z", // Correct!
    "EndDate": "2018-08-31T23:59:00Z",
    ...
}
```

[Related StackOverflow question here](https://stackoverflow.com/questions/52561304/all-day-event-dates-incorrect-in-calendar-listitems-created-in-graph-api)

**Comment by jxnx888** [comment]
Hello @mikepowell , could you please tell me what's your language of this code?

**Comment by mikepowell** [comment]
The code in the issue text is portions of request and response bodies in JSON format. (I tagged the code blocks as JavaScript because I wanted to add comments, which aren't technically supported in JSON.)

As far as the issue itself I discovered it in C# code that uses the .NET Graph API, but have reproduced the issue using raw REST requests in Graph Explorer.

**Comment by jxnx888** [comment]
For now, one of the workarounds is changing the "fAllDayEvent" to "false".

**Comment by mikepowell** [comment]
Yes, that works but does not create an all-day event of course. My requirement is to have these events displayed in SharePoint and Outlook as all-day events rather than events with start and end times.

**Comment by tigertop** [comment]
I can confirm that same behavior on our tenant, but it works correctly on our client's tenant. Our tenant was created like a year ago and is on the fast track. Client's tenant was created like 1.5 months ago.

**Comment by tvkit** [comment]
Did you try specifying midnight as the end point?

For example, in C# ?

```
var start = when.ToUniversalTime().Date;
var end = start.AddDays(1);

```
The delta on `(end - start)` is one day, not one day plus a nanosecond for example. So I'd expect this to work for an all-day event.

When specifying 11:59 PM as the end point, the All Day calc may be subtracting 24 hours from the end point going back to the previous day. Just a thought.

**Comment by svwerner** [comment]
The same problem appeared today then creating a new calendar event via MS Graph API. Now 2 days are beeing substracted from the EventDate submitted. Disableing the all day event resultet in a substraction of 1 day of the submitted date.

The request has been made with an app registration using the Microsoft Graph API Endpoint for SharePoint Online.


```
POST: https://graph.microsoft.com/v1.0/sites/{sitename}/lists/{listId}/items
```

Body
```json
{
   "fields": {
      "ContentType": "Ereignis",
      "Title": "SharePoint Date Bug",
      "EventDate": "2020-01-03T00:00:00Z",
      "EndDate": "2020-01-12T23:59:00Z",
      "Description": "",
      "fAllDayEvent": true,
      "fRecurrence": false
    }
}
```

Response ( .. only the field data is shown here)
```json
{
    "fields": {
        "@odata.etag": "\"97070c93-8810-4c6c-b86c-61df55d12a0e,1\"",
        "id": "90",
        "ContentType": "Ereignis",
        "Title": "SharePoint Date Bug",
        "Modified": "2020-01-10T05:17:57Z",
        "Created": "2020-01-10T05:17:57Z",
        "AuthorLookupId": "1073741822",
        "EditorLookupId": "1073741822",
        "_UIVersionString": "1.0",
        "Attachments": false,
        "Edit": "",
        "LinkTitleNoMenu": "SharePoint Date Bug",
        "LinkTitle": "SharePoint Date Bug",
        "ItemChildCount": "0",
        "FolderChildCount": "0",
        "_ComplianceFlags": "",
        "_ComplianceTag": "",
        "_ComplianceTagWrittenTime": "",
        "_ComplianceTagUserId": "",
        "AppAuthorLookupId": "24",
        "AppEditorLookupId": "24",
        "EventDate": "2020-01-01T00:00:00Z",
        "EndDate": "2020-01-12T23:59:00Z",
        "fAllDayEvent": true,
        "fRecurrence": false
}
```

Note that the response event date does not match the request previously sent.

Request = ` "EventDate": "2020-01-03T00:00:00Z"`
Response = `"EventDate": "2020-01-01T00:00:00Z"`

Adding days before sending the request to SharePoint cannot be a viable long term solution. 
A fix for this problem would be highly appreciated!

**Comment by Stezii** [comment]
I am facing a 1-day shift. When and how does this vary? Do You have any news on this bug?

**Comment by majoshimsft** [comment]
Passing body like below, should fix it.
{
    "fields": {
        "Title": "Test11",
        "EventDate": "2021-10-28T23:59:00Z",
        "EndDate": "2021-10-28T00:00:00Z",
        "fAllDayEvent": true,
        "fRecurrence": false
    }
}

**Comment by dospencer** [comment]
Thank you @majoshimsft 

Is that a workaround or is that how it should work? It doesn’t seem logical that I have to make the EventDate (start date) greater than the EndDate. My fear is that when this “bug” is fixed, my calendar dates will be thrown off. 

I tried the suggestion nonetheless and instead of creating the event for the date specified, it created it for a day earlier. So, in the example body below, the event was created for 10/29 instead of 10/30 as specified with the dates. Incidentally, MS Support made the same suggestion but I asked them to investigate the issue some more. 


"{
    \"fields\": {
        \"Title\": \"PTO - Test 4\",
        \"EventDate\": \"2021-10-30T23:59:00Z\",
        \"EndDate\": \"2021-10-30T00:00:00Z\",
        \"fAllDayEvent\": true,
        \"fRecurrence\": false,
        \"Category\": \"Vacation Day\"
    }
}"

**Comment by dospencer** [comment]
I opened a support case with Microsoft and they indicated that the API was not correctly handling some timezone conversions. Their engineering team has fixed the issue and the fix is expected to roll out completely by end of the month.

**Comment by dospencer** [comment]
Microsoft has delivered a fix. I tested it and it works. I can now specify the exact start and end date for my all day event and they populate the calendar correctly.
